applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -137,36 +147,72 @@ static void handle_dvfs_error(int32_t err)
 		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
 	} else {
 		LOG_ERR("DVFS freq change returned with error: %d", err);
-		module_set_state(MODULE_STATE_ERROR);
-		module_state = STATE_ERROR;
-		cancel_dvfs_retry_work();
+		dvfs_fatal_error();
 		return;
 	}
 	(void) k_work_reschedule(&dvfs_retry.retry_work, timeout);
 }
 
-static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+static void dvfs_notify_cb(void)
 {
-	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+	(void) k_work_submit(&dvfs_notify_work);
+}
+
+static void dvfs_notify_work_handler(struct k_work *work)
+{
+	int res;
+	int32_t ret = sys_notify_fetch_result(&cli.notify, &res);
+
+	if (ret < 0) {
+		LOG_DBG("Work not completed");
+		return;
+	}
+
+	sys_notify_init_callback(&cli.notify, dvfs_notify_cb);
+
+	if (res < 0) {
+		handle_dvfs_error(res);
+		return;
+	}
+	ret = clock_control_get_rate(dvfs_clock_dev, 0, &current_freq);
+	if (ret < 0) {
+		LOG_ERR("Failed to get current frequency");
+		dvfs_fatal_error();
+		return;
+	}
+	LOG_INF("DVFS completed, current frequency is: %u", current_freq);
+}
+
+static void set_dvfs_freq(uint32_t target_freq)
+{
+	int32_t ret = nrf_clock_control_release(dvfs_clock_dev, &spec);
+
+	if (ret == -ENOTSUP) {
+		LOG_DBG("No supported clock control to be released");
+	} else if (ret < 0) {
+		LOG_ERR("Failed to release requested clock specs");
+		dvfs_fatal_error();
+		return;
+	}
 
+	spec.frequency = target_freq;
+	ret = nrf_clock_control_request(dvfs_clock_dev, &spec, &cli);
 	if (ret) {
 		handle_dvfs_error(ret);
 	} else {
-		current_freq = target_freq;
-		LOG_INF("Have requested %s frequency",
-			get_dvfs_frequency_setting_name(target_freq));
+		LOG_INF("Have requested %u frequency", target_freq);
 		cancel_dvfs_retry_work();
 	}
 }
 
-static enum dvfs_frequency_setting check_required_frequency(void)
+static uint32_t check_required_frequency(void)
 {
 	if (dfvs_requests_state_bitmask & dvfs_high_freq_bitmask) {
```

Comment:
don't understand what you mean here.

Reply:
I quite like the idea, but I see a problem that we don't have a bitmask for low_frequency.

Because of that I find this solution more confusing then current situation. If you have idea how to solve it feel free to comment :)

Reply:
You can e.g. put the low frequency outside of the array of structs (if nothing other is requested you fallback to the low frequency that is used as a default)

Reply:
We have all of the "increased" frequencies in one place (only the default/lowest one would be left out). That seems reasonable and would allow us to keep implementation easier to expand. @mkapala-nordic what do you think?

Still, this is only an improvement suggestion and might be done as a follow-up action (or ignored if we decide it's not worth it).

Reply:
Better idea would be to create also struct object for lowest frequency and fill bitmask with "1" that would be more consistent.

For me 5 variables ~ 3 structs >>>  2 structs + 1 variable

Reply:
Array of 3 structs will be the best IMO :D If not, then leave as is, I would not mix the approaches.

Reply:
Then the last struct would need to have bitmask of all `1`? Sounds reasonable too :)

Reply:
IMHO it would make implementation a bit cleaner, but it's not a must have (we can e.g. do it as a follow-up action)

Reply:
ok introduced change